File: lang/python/src/fun/vac/data_structures/ArrayQueue.py
class ArrayQueue:

    DEFAULT_CAPACITY = 64

    def __init__(self, physicalSize: int = 0):
        self.data = None
        self.front = 0
        self.logicalSize = 0
        self.physicalSize = self.DEFAULT_CAPACITY

        if physicalSize > 1:
            self.physicalSize = physicalSize
        self.data = [0] * self.physicalSize

    def size(self) -> int:
        return self.logicalSize

    def isEmpty(self) -> bool:
        return self.logicalSize == 0

    def peek(self) -> int:
        if self.logicalSize == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        return self.data[self.front]

    def enqueue(self, element: object) -> None:
        if self.logicalSize == self.physicalSize:
            newCapacity = self.DEFAULT_CAPACITY

            if self.physicalSize >= self.DEFAULT_CAPACITY:
                newCapacity = self.physicalSize + (self.physicalSize >> 1)

            temp = [0] * newCapacity

            cursor = self.front

            for i in range(self.logicalSize):
                if cursor == self.physicalSize:
                    cursor = 0
                temp[i] = self.data[cursor]
                cursor += 1

            self.data = temp
            self.front = 0
            self.physicalSize = newCapacity

        self.data[(self.front + self.logicalSize) % self.physicalSize] = element

        self.logicalSize += 1

    def dequeue(self) -> None:
        if self.logicalSize == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        self.data[self.front] = None

        self.front = (self.front + 1) % self.physicalSize

        self.logicalSize -= 1

    def capacity(self) -> int:
        return self.physicalSize

    def shrink(self) -> None:
        temp = [0] * self.logicalSize

        cursor = self.front

        for i in range(self.logicalSize):
            if cursor == self.physicalSize:
                cursor = 0
            temp[i] = self.data[cursor]
            cursor += 1

        self.data = temp
        self.front = 0
        self.physicalSize = self.logicalSize

File: lang/ruby/src/fun/vac/data_structures/ArrayQueue.rb
module ArrayQueue

    class ArrayQueue

        @@DEFAULT_CAPACITY = 64

        public
        def initialize(physicalSize = 0)
            @data = nil
            @front = 0
            @logicalSize = 0
            @physicalSize = @@DEFAULT_CAPACITY

            if physicalSize > 1
                @physicalSize = physicalSize
            end
            @data = Array.new(@physicalSize)
        end

        public
        def size()
            @logicalSize
        end

        def isEmpty()
            @logicalSize.zero?
        end

        def peek()
            if @logicalSize.zero?
                raise "[PANIC - NoSuchElement]"
            end

            @data[@front]
        end

        def enqueue(element)
            if @logicalSize == @physicalSize
                newCapacity = @@DEFAULT_CAPACITY

                if @physicalSize >= @@DEFAULT_CAPACITY
                    newCapacity = @physicalSize + (@physicalSize >> 1)
                end

                temp = Array.new(newCapacity)

                cursor = @front

                0.upto(@logicalSize - 1) do |i|
                    if cursor == @physicalSize
                        cursor = 0
                    end
                    temp[i] = @data[cursor]
                    cursor += 1
                end

                @data = temp
                @front = 0
                @physicalSize = newCapacity
            end

            @data[(@front + @logicalSize) % @physicalSize] = element

            @logicalSize += 1
        end

        def dequeue()
            if @logicalSize.zero?
                raise "[PANIC - NoSuchElement]"
            end

            @data[@front] = nil

            @front = (@front + 1) % @physicalSize

            @logicalSize -= 1
        end

        def capacity()
            @physicalSize
        end

        def shrink()
            temp = Array.new(@logicalSize)

            cursor = @front

            0.upto(@logicalSize - 1) do |i|
                if cursor == @physicalSize
                    cursor = 0
                end
                temp[i] = @data[cursor]
                cursor += 1
            end

            @data = temp
            @front = 0
            @physicalSize = @logicalSize
        end
    end
end
